File: scrapers/icici_scraper.py
from scrapers.base_scraper import BaseScraper
from core.extractor import LayeredExtractor
from typing import Dict, Any

class ICICIScraper(BaseScraper):
    """
    Scraper implementation for ICICI Bank.
    """
    async def scrape(self, page: Any) -> Dict[str, Any]:
        self.logger.info("starting_icici_scrape")
        
        if self.url.lower().endswith(".pdf"):
            import tempfile
            pdf_dir = tempfile.gettempdir()
            response = await page.context.request.get(self.url)
            temp_pdf = f"{pdf_dir}/icici_rates.pdf"
            with open(temp_pdf, "wb") as f:
                f.write(await response.body())
            rates = LayeredExtractor.extract_from_pdf(temp_pdf)
        else:
            try:
                await page.wait_for_selector("table, [role='table']", timeout=5000)
            except Exception:
                self.logger.warning("timeout_waiting_for_tables_attempting_anyway")
            
            tables = await LayeredExtractor.extract_from_page(page)
            rates = []
            for t in tables:
                parsed = LayeredExtractor.parse_extracted_table(t)
                if parsed:
                    rates.extend(parsed)
            
            if not rates:
                rates = await LayeredExtractor.extract_from_unstructured_text(page)
                
        return {
            "fd_rates": rates,
            "minimum_deposit": 10000.0,
            "maximum_deposit": 20000000.0,
            "premature_withdrawal_available": True,
            "premature_withdrawal_penalty": "0.50% to 1.00% depending on tenure and deposit value",
            "loan_against_fd_available": True,
            "tax_saver_fd_available": True,
            "tax_saver_tenure": "5 Years",
            "nomination_available": True,
            "compounding_frequency": "Quarterly",
            "last_updated_on_page": None,
            "effective_from": None,
            "effective_to": None,
            "scraper_version": "1.0.0"
        }

File: scrapers/iciciScraper.js
import fs from 'fs';
import path from 'path';
import os from 'os';
import { BaseScraper } from './baseScraper.js';
import { LayeredExtractor } from '../core/extractor.js';

export class ICICIScraper extends BaseScraper {
  async scrape(page) {
    let rates = [];

    if (this.url.toLowerCase().endsWith(".pdf")) {
      const tempPdf = path.join(os.tmpdir(), "icici_rates.pdf");
      const response = await page.context().request().get(this.url);
      fs.writeFileSync(tempPdf, await response.body());
      rates = await LayeredExtractor.extractFromPdf(tempPdf);
    } else {
      try {
        await page.waitForSelector("table, [role='table']", { timeout: 5000 });
      } catch (e) {
      }

      // Click "Less than 3 Cr." button if it exists
      try {
        const btn = page.locator("button:has-text('Less than')").filter({ hasText: "3 Cr" });
        if (await btn.count() > 0) {
          await btn.first().click();
          await page.waitForTimeout(2000);
        }
      } catch (e) {
      }

      const tables = await LayeredExtractor.extractFromPage(page);
      for (const t of tables) {
        const parsed = LayeredExtractor.parseExtractedTable(t);
        if (parsed && parsed.length > 0) {
          rates.push(...parsed);
        }
      }

      if (rates.length === 0) {
        rates = await LayeredExtractor.extractFromUnstructuredText(page);
      }
    }

    return {
      fd_rates: rates,
      minimum_deposit: 10000.0,
      maximum_deposit: 20000000.0,
      premature_withdrawal_available: true,
      premature_withdrawal_penalty: "0.50% to 1.00% depending on tenure and deposit value",
      loan_against_fd_available: true,
      tax_saver_fd_available: true,
      tax_saver_tenure: "5 Years",
      nomination_available: true,
      compounding_frequency: "Quarterly",
      last_updated_on_page: null,
      effective_from: null,
      effective_to: null,
      scraper_version: "1.0.0"
    };
  }
}
